### PR TITLE
Add Voxtral Mini 3B contrib model (audio-language)

### PR DESCRIPTION
**Pull request by jimburtoft**
*Note: The below template includes items meant for model contributions only. For other contributions such as bug fixes, features, etc., only fill out the relevant portions of the form.*

## Description

Adds Voxtral-Mini-3B (`mistralai/Voxtral-Mini-3B-2507`), Mistral AI's audio-language model, as a contrib model in `contrib/models/voxtral-mini-3B/`.

Voxtral is decomposed into three components on Neuron:
- **Audio encoder** (Whisper-derived, 637 M params, 32 layers) traced via `torch_neuronx.trace()`.
- **Projector** (25 M params, 4-frame packing Linear→GELU→Linear) running on CPU.
- **Text decoder** (Ministral-3B, 3.3 B params, 30 layers, GQA 32/8) via NxDI's `NeuronLlamaModel` with scatter-based audio-token injection through the existing `scatter_by_index_put` used by Pixtral / Llama-4 vision.

The whole pipeline is orchestrated by a top-level `NeuronApplicationVoxtral` in `src/modeling_voxtral.py` that handles text-only weight extraction, encoder trace + `move_trace_to_device` load, projector CPU load, and NxDI text-decoder compile/load.

Both text-only chat and audio transcription are supported end-to-end. Verified on trn2.3xlarge SDK 2.30 (DLAMI 20260522) at TP=4, LNC=2, bfloat16.

## Model Information

**Model Name:** Voxtral-Mini-3B-2507 (`mistralai/Voxtral-Mini-3B-2507`)

**Model Architecture:** Audio-language multimodal (Whisper-style audio encoder + linear projector + Llama decoder). Encoder-decoder for audio inputs, decoder-only for text-only.

**Purpose:** Speech transcription and audio understanding (English + multilingual). Up to 30 seconds of audio per request.

## Checklist

- [x] **Accuracy Test** (`test/integration/test_model.py`)
  - Uses `check_accuracy_logits_v2` for text-only logit-level validation against CPU BF16 reference (16 tokens, multi-tier tolerances).
  - Uses direct token-ID comparison against CPU BF16 reference for the audio-transcription path (loose match tolerance of one drift token allowed).
  - Skippable with `VOXTRAL_SKIP_CPU_REF=1` for smoke-only runs.
  - Test can compile and run the model on Neuron.

- [x] **README.md**
  - [x] **Usage Example**: Both standalone `NeuronApplicationVoxtral.transcribe()` and a pointer to the vLLM walkthrough.
  - [x] **Compatibility Matrix**: trn2.3xlarge on SDK 2.28/2.30/2.31, inf2.xlarge on SDK 2.28.
  - [x] **Example Checkpoints**: `mistralai/Voxtral-Mini-3B-2507` on HuggingFace.
  - [x] **Testing Instructions**: pytest invocation with required env vars.

- [x] **Source Code** (`src/`)
  - `modeling_voxtral.py` with `NeuronApplicationVoxtral`, `VoxtralForCausalLM` (NxDI `NeuronBaseForImageToText`), `VoxtralTextModel` (extends `NeuronLlamaModel` with scatter injection), `VoxtralInferenceConfig` (builds a `PixtralInferenceConfig`).
  - `src/__init__.py` exports the top-level classes.
  - Follows contrib folder hierarchy.

### Optional Components

- [x] **Benchmark harness** (`benchmark_harness/`): serial latency driver with `nxdi_neuron` and `vllm_neuron` backends, empty dataset skeleton for customer-supplied audio, summarise script.
- [x] **Walkthrough notebook** (`voxtral_trn2_walkthrough.ipynb`): end-to-end guide for trn2.3xlarge SDK 2.30, with cell outputs preserved from a live execution.
- [x] **Encoder benchmark** (`benchmark_encoder.py`): component-level latency measurement (retained from initial contrib).

## Folder Structure

```
/contrib/models/voxtral-mini-3B/
├── README.md
├── voxtral_trn2_walkthrough.ipynb
├── benchmark_encoder.py
├── benchmark_harness/
│   ├── README.md
│   ├── run_voxtral_benchmark.py
│   ├── backends.py
│   ├── backend_nxdi.py
│   ├── backend_vllm.py
│   ├── summarize_results.py
│   ├── common/{__init__.py, audio.py, manifest.py, timing.py}
│   └── dataset/{README.md, manifest.csv}
├── src/
│   ├── __init__.py
│   ├── modeling_voxtral.py
│   └── utils/__init__.py
└── test/
    ├── __init__.py
    └── integration/
        ├── __init__.py
        └── test_model.py
```

## Testing

**How did you test this change?**

End-to-end on trn2.3xlarge, DLAMI `Deep Learning AMI Neuron (Ubuntu 24.04) 20260522` (Neuron SDK 2.30), venv `/opt/aws_neuronx_venv_pytorch_inference_vllm_0_16/`. Two paths tested:

1. **Standalone `NeuronApplicationVoxtral.transcribe()`** with `tp_degree=4, seq_len=512, n_positions=768, on_device_sampling=True, move_trace_to_device=True, bfloat16`. Mean latency 0.477 s per file on a 5-clip 5-25 s TED-style harness (3 runs, 15 measurements).

2. **vLLM-neuron serving** via `LLM(...)` with `additional_config={"override_neuron_config": {...}}` and `seq_len=n_positions=768`. Mean latency 0.489 s per file on the same harness (from the notebook execution).

The full walkthrough notebook was executed via `jupyter nbconvert --to notebook --execute --inplace` on the trn2 instance and the outputs are committed.

**Test Results:**

Per-duration bin means (ms), 5-clip harness:

| Duration | Standalone | vLLM |
|----------|:---------:|:----:|
| 5-10 s   | 184       | 288  |
| 10-15 s  | 419       | 400  |
| 15-20 s  | 590       | 436  |
| 20-25 s  | 700       | 607  |
| 25-30 s  | --        | 714  |

Text-only logit validation via `check_accuracy_logits_v2` on 16 tokens: PASS with default tolerances (`divergence_difference_tol=0.02`).

Audio transcription token match vs CPU BF16 reference: 15/16 tokens matched on the TED-60 sample (one drift token allowed by the test).

## Compatibility

**Tested with:**
- **Neuron SDK Version(s):** 2.28, 2.30, 2.31 (best results on 2.30)
- **Instance Type(s):** trn2.3xlarge (primary), inf2.xlarge
- **PyTorch Version:** 2.9.0 (SDK 2.30)
- **Python Version:** 3.12

Both `neuronx-distributed-inference` 0.10.17970 (SDK 2.30) and 0.10.18399 (SDK 2.31) are supported. NxDI 0.9.x may work but has not been verified in this PR.

### Optimizations shipped

Four changes over the initial port bring per-file latency from 0.656 s (stock SDK 2.31, TP=4) to ~0.47-0.50 s (SDK 2.30):

1. **`torch_neuronx.move_trace_to_device(encoder, 0)`** at load time -- removes ~95 ms of per-call host overhead on the audio encoder. Default `True`; fall back to `async_load` via constructor arg.
2. **On-device sampling** (`on_device_sampling=True`) -- greedy argmax runs inside the NEFF via `OnDeviceSamplingConfig(top_k=1, do_sample=False, dynamic=False, deterministic=False)`. Both compile and runtime must agree.
3. **`n_positions=768`** (default `4096`) -- shrinks the KV cache to fit ~375 audio tokens + short prompt.
4. **`seq_len=512`** (CTE bucket, default `2048`) -- matches actual audio prefill length. Standalone path only; see Known Limitations for the vLLM constraint.

### Known limitations

- **Batch size 1 only.** `scatter_by_index_put` in the stock `ImageToTextModelWrapper` is not shape-safe for BS>1 CTE.
- **30 s audio maximum per request.** Voxtral upstream's 30 min / 40 min long-form modes are out of scope.
- **vLLM V1 requires `seq_len == n_positions`.** The V1 scheduler feeds the TKG NEFF positions in `[0, n_positions)` during decode; with `seq_len < n_positions` the TKG NEFF rejects positions in the gap with NRT status 1006. Standalone path is not affected (it pads inputs to `seq_len` internally).
- **`--auto-cast=matmult`** is applied on the encoder trace only (required for the FP32→BF16 path in Whisper attention). NxDI handles the LLM path.

## vLLM Integration

- [x] This model is intended for use with vLLM
- [x] Documentation includes vLLM registration instructions

The walkthrough notebook (`voxtral_trn2_walkthrough.ipynb`) demonstrates the full vLLM-neuron flow: cloning the [jimburtoft/vllm-neuron `contrib/voxtral-0.5.0`](https://github.com/jimburtoft/vllm-neuron/tree/contrib/voxtral-0.5.0) fork, installing it in the DLAMI vLLM venv, and serving the model via `vllm.LLM(...)` with `additional_config={"override_neuron_config": {"on_device_sampling": True, "move_trace_to_device": True, "n_positions": 768, "seq_len": 768}}`.

The vllm-neuron fork adds a `NeuronVoxtralForCausalLM` class that wraps `NeuronApplicationVoxtral` and handles vLLM 0.16.0's `audio_arrays` multi-modal kwargs (raw 16 kHz waveforms), converting them to log-mel spectrograms via HF's `WhisperFeatureExtractor` on the CPU per request.

*For vLLM integration details, see: https://awsdocs-neuron.readthedocs-hosted.com/en/latest/libraries/nxd-inference/developer_guides/onboarding-models.html#nxdi-onboarding-models-vllm*

---

**By submitting this PR, I confirm that:**
- [x] I have read and followed the [contributing guidelines](../contrib/CONTRIBUTING.md)
- [x] This is a community contribution and may have limited testing compared to officially-supported models
- [x] The code follows best practices and is well-documented
- [x] All required components listed above are included

### CODE REVIEW

**Comment by jimburtoft** [comment]
**2026-07-30 update**: Refreshed with latency optimizations and a working vLLM-neuron serving path.

New in this iteration:

1. `NeuronApplicationVoxtral` now accepts `on_device_sampling` (default False) and `move_trace_to_device` (default True) kwargs. Combined with tuning `n_positions=768` and `seq_len=512`, the standalone customer-harness mean drops from 0.656 s/file (stock SDK 2.31) to 0.477 s/file (SDK 2.30, 5-clip 5-25 s TED-style harness).
2. Text-only integration test now uses `check_accuracy_logits_v2` for logit-level validation; audio test compares Neuron token IDs to a CPU BF16 reference (both replace the earlier loose `"paris" in output.lower()` check).
3. Added `benchmark_harness/` -- a serial per-file latency driver with `nxdi_neuron` and `vllm_neuron` backends and an empty dataset skeleton for user-supplied audio.
4. Added `voxtral_trn2_walkthrough.ipynb` -- end-to-end walkthrough for trn2.3xlarge SDK 2.30, executed on hardware with outputs preserved. Notebook run reported 489 ms mean per file across the 5-clip harness (3 runs, 15 measurements).
5. Rebased on top of the current upstream `main`.

Companion vllm-neuron patch: https://github.com/jimburtoft/vllm-neuron/tree/contrib/voxtral-0.5.0 (contains an `audio_arrays` handler for vLLM 0.16.0's Voxtral processor plus a vision-mask padding fix).

Verified on trn2.3xlarge in ap-southeast-4 today.

**Comment by jimburtoft** [comment]
**2026-07-30 verification**: The walkthrough notebook has been re-executed end-to-end on a **freshly provisioned trn2.3xlarge** with no pre-cached model, no pre-cached compile, no cloned forks, and no HF cache. Every step that could be a no-op on a warm host actually ran:

- Cell 6: real `git clone` of both forks
- Cell 8: real `pip install -e` of the vllm-neuron fork (replacing the DLAMI's pre-installed vllm-neuron 0.5.0)
- Cell 14: real 18 GB HuggingFace snapshot download
- Cell 18: real Neuron compile of the encoder + decoder NEFFs
- Cell 20: real vLLM engine launch against the newly-compiled artefacts
- Cell 23: real single-file transcription smoke test
- Cell 26: real 3-run, 15-measurement harness benchmark
- Cell 27: **`Mean latency: 501.9 ms, median 458.9 ms`** on the 5-clip auto-generated harness -- under the 543 ms per-file target

The committed notebook has cell outputs captured from that run. The only code cell without captured output is cell 12 (`pip install --quiet`) which produced no stdout because all packages were already satisfied in the DLAMI's pre-installed `/opt/aws_neuronx_venv_pytorch_inference_vllm_0_16/`.

Also cleaned up all remaining "customer" phrasing per the contrib CONTRIBUTING.md prohibition on customer references (there were 5 uses of "customer-facing / customer-supplied / customer benchmark harness" that were generic but read as if this contribution was motivated by a specific customer engagement). The submission is now zero for {"customer", agent names, internal project names, cost info, GPU comparisons, specific compile times}.

**Comment by jimburtoft** [comment]
**2026-07-30 SDK 2.31 verification**: Also re-tested the notebook top-to-bottom on a **freshly provisioned trn2.3xlarge** with the SDK 2.31 DLAMI (`Deep Learning AMI Neuron (Ubuntu 24.04) 20260708`, AMI `ami-0b658f2749710f886` in ap-southeast-4).

Stack:

- `neuronx-cc` 2.26.6360
- `neuronx-distributed-inference` 0.10.18399
- `torch-neuronx` 2.9.0.2.15.32035
- `vllm` 0.16.0 + `vllm-neuron` 0.5.0 (patched)

Result: **notebook executes cleanly, 14/15 code cells have output, 0 errors**. Harness reports:

```
N files:        15
Mean latency:   502.2 ms
Median latency: 458.1 ms
P90 latency:    731.2 ms
P99 latency:    734.4 ms
```

Effectively identical to the SDK 2.30 run (mean 501.9 ms). Under the 543 ms per-file target on both SDK 2.30 and SDK 2.31, so the compatibility matrix's "Validated" claim for both rows is backed by an actual end-to-end run in this PR.
